src/fd.cpp

```diff
@@ -0,0 +1,248 @@
+#ifdef _WIN32
+#ifndef FD_SETSIZE
+#define FD_SETSIZE 2048  // Set max fds - only for select() fallback
+#endif
+#include <winsock2.h>
+#else
+#include <poll.h>
+#endif
+#include <Rcpp.h>
+#include <unistd.h>
+#include <cstdlib>
+#include <atomic>
+#include "tinycthread.h"
+#include "later.h"
+#include "callback_registry_table.h"
+
+#if R_VERSION < R_Version(4, 5, 0)
+inline SEXP R_mkClosure(SEXP formals, SEXP body, SEXP env) {
+  SEXP fun = Rf_allocSExp(CLOSXP);
+  SET_FORMALS(fun, formals);
+  SET_BODY(fun, body);
+  SET_CLOENV(fun, env);
+  return fun;
+}
+#endif
+
+extern CallbackRegistryTable callbackRegistryTable;
+extern SEXP later_fdcancel;
+extern SEXP later_invisibleSymbol;
+
+#define LATER_INTERVAL 1024
+
+#ifdef _WIN32
+#define POLL_FUNC WSAPoll
+#else
+#define POLL_FUNC poll
+#endif
+
+typedef struct ThreadArgs_s {
+  std::shared_ptr<std::atomic<bool>> flag;
+  std::unique_ptr<std::vector<int>> fds;
```

Comment:
What's the purpose of this? It looks to me like you initialize it to contain all the fd's, but then ignore it, then use it as a `int*` to hold result values. Can you get rid of it and create a new `Rcpp::LogicalVector` on the stack at callback time?

Reply:
Sorry, I see now that it is the only place you hold the fd's. I find it confusing that it's used as both input and output. Can I suggest that you keep it to just input?

Reply:
OK, I've just had to bite the bullet here and introduce a results vector for the args struct in [a32b3ea](https://github.com/r-lib/later/pull/190/commits/a32b3ea84c9596b528508b61dd538c61ae06b995).

Although not strictly necessary, it does make things clearer and I didn't find a really good alternative that wasn't confusing in some other way.

I note it's a bit like the separate 'events' and 'revents' fields of `poll()` rather than a self-updating structure like the `select()` 'fd_sets', with `poll()` being the more modern design.